src/main/java/com/artipie/api/artifactory/AddUpdateUserSlice.java

```diff
@@ -68,26 +66,23 @@ public AddUpdateUserSlice(final Settings settings) {
     @Override
     public Response response(final String line, final Iterable<Map.Entry<String, String>> headers,
         final Publisher<ByteBuffer> body) {
-        final Response res;
-        final Matcher matcher = GetUserSlice.PTRN.matcher(
-            new RequestLineFrom(line).uri().toString()
-        );
-        if (matcher.matches()) {
-            final CompletionStage<Optional<String>> pswd = AddUpdateUserSlice.password(body);
-            final String username = matcher.group("username");
-            res = new AsyncResponse(
-                pswd.thenCompose(
-                    passw -> passw.map(
-                        haspassw -> this.settings.credentials()
-                            .thenCompose(
-                                cred -> cred.add(username, haspassw)
-                                    .thenApply(ok -> new RsWithStatus(RsStatus.OK)))
+        final Optional<String> user = new UserFromRqLine(line).get();
+        return user.map(
+            username -> new AsyncResponse(
+                AddUpdateUserSlice.password(body)
+                    .thenCompose(
+                        passw -> passw.map(
+                            haspassw -> this.settings.credentials()
+                                .thenCompose(
+                                    cred -> cred.add(username, haspassw)
+                                        .thenApply(ok -> new RsWithStatus(RsStatus.OK)))
                 ).orElse(CompletableFuture.completedFuture(new RsWithStatus(RsStatus.NOT_FOUND))))
-            );
-        } else {
-            res = new RsWithStatus(RsStatus.BAD_REQUEST);
-        }
-        return res;
+            )
+        ).orElse(
+            new AsyncResponse(
+                CompletableFuture.completedFuture(new RsWithStatus(RsStatus.BAD_REQUEST))
+            )
```

Comment:
@genryxy you may return `new RsWithStatus(RsStatus.BAD_REQUEST)` without wrapping it into `CompletableFuture` and then into `AsyncResponse`.

Reply:
@olegmoz thanks, sorry for the delaying. If I remove both wrappers, it doesn't compile.

Reply:
@genryxy probably you need to specify `<Response>` type explicitly for `Optional.map` method call. Otherwise it is resoled automatically to `AsyncResponse` and then `AsyncResponse` is expected to be returned in `orElse`

Reply:
@olegmoz yes, you're right, my bad